### PR TITLE
[Bug] Fix substitute interactions with `PostDefendAbAttr`s

### PR DESCRIPTION
**Pull request by DayKev**
## What are the changes the user will see?
The abilities will no longer incorrectly trigger when the enemy's attack hits the user's substitute.

## Why am I making these changes?
The behavior is incorrect.

## What are the changes from a developer perspective?
Added a `hitsSubstitute()` check to the `PostDefendAbAttr`s.

## How to test the changes?
Added tests:
`npm run test disguise`
`npm run test ice_face`
`npm run test gulp_missile`

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- ~[ ] If I have text, did I make it translatable and add a key in the English locale file(s)?~
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- ~[ ] Are the changes visual?~
  - ~[ ] Have I provided screenshots/videos of the changes?~

### CODE REVIEW

**Comment by Snailman11** [comment]
Link to Discord's Report (Mimikyu's Disguise gets hit before Substitute)
https://discord.com/channels/1125469663833370665/1291084587643961455

**Review by innerthunder — APPROVED** [review]

File: src/test/abilities/disguise.test.ts [hunk 1]
```diff
@@ -222,4 +223,17 @@ describe("Abilities - Disguise", () => {
...
   });
+
+  it("doesn't trigger if user is behind a substitute", async () => {
+    game.override
+      .enemyMoveset(Moves.SUBSTITUTE)
+      .moveset(Moves.POWER_TRIP);
+    await game.classicMode.startBattle();
+
+    game.move.select(Moves.POWER_TRIP);
+    await game.setTurnOrder([BattlerIndex.ENEMY, BattlerIndex.PLAYER]);
```
Comment (DayKev) [inline src/test/abilities/disguise.test.ts:234]:
```suggestion
    await game.setTurnOrder([ BattlerIndex.ENEMY, BattlerIndex.PLAYER ]);
```

File: src/test/abilities/gulp_missile.test.ts [hunk 2]
```diff
@@ -234,6 +236,25 @@ describe("Abilities - Gulp Missile", () => {
     expect(game.scene.getEnemyPokemon()!.getStatStage(Stat.DEF)).toBe(-1);
   });
 
+  it("doesn't trigger if user is behind a substitute", async () => {
+    game.override
+      .enemyAbility(Abilities.STURDY)
+      .enemyMoveset([Moves.SPLASH, Moves.POWER_TRIP]);
```
Comment (DayKev) [inline src/test/abilities/gulp_missile.test.ts:242]:
```suggestion
      .enemyMoveset([ Moves.SPLASH, Moves.POWER_TRIP ]);
```

File: src/test/abilities/gulp_missile.test.ts [hunk 3]
```diff
@@ -234,6 +236,25 @@ describe("Abilities - Gulp Missile", () => {
...
+    await game.classicMode.startBattle([Species.CRAMORANT]);
```
Comment (DayKev) [inline src/test/abilities/gulp_missile.test.ts:243]:
```suggestion
    await game.classicMode.startBattle([ Species.CRAMORANT ]);
```

File: src/test/abilities/gulp_missile.test.ts [hunk 4]
```diff
@@ -234,6 +236,25 @@ describe("Abilities - Gulp Missile", () => {
...
+
+    game.move.select(Moves.SURF);
+    await game.forceEnemyMove(Moves.SPLASH);
+    await game.toNextTurn();
+
+    expect(game.scene.getPlayerPokemon()!.formIndex).toBe(GULPING_FORM);
+
+    game.move.select(Moves.SUBSTITUTE);
+    await game.forceEnemyMove(Moves.POWER_TRIP);
+    await game.setTurnOrder([BattlerIndex.PLAYER, BattlerIndex.ENEMY]);
```
Comment (DayKev) [inline src/test/abilities/gulp_missile.test.ts:253]:
```suggestion
    await game.setTurnOrder([ BattlerIndex.PLAYER, BattlerIndex.ENEMY ]);
```

File: src/test/abilities/ice_face.test.ts [hunk 5]
```diff
@@ -222,10 +223,23 @@ describe("Abilities - Ice Face", () => {
...
   });
 
+  it("doesn't trigger if user is behind a substitute", async () => {
+    game.override
+      .enemyMoveset(Moves.SUBSTITUTE)
+      .moveset(Moves.POWER_TRIP);
+    await game.classicMode.startBattle();
+
+    game.move.select(Moves.POWER_TRIP);
+    await game.setTurnOrder([BattlerIndex.ENEMY, BattlerIndex.PLAYER]);
```
Comment (DayKev) [inline src/test/abilities/ice_face.test.ts:233]:
```suggestion
    await game.setTurnOrder([ BattlerIndex.ENEMY, BattlerIndex.PLAYER ]);
```

**Review by innerthunder — DISMISSED** [review]

**Review by innerthunder — COMMENTED** [review]
The cleanup of `MoveEffectPhase` should probably be put into another PR

**Comment by DayKev** [comment]
Oh yeah, I originally had some changes in that file but swapped to a different solution later so all that's left is the formatting changes. Lemme pull the formatting out.

**Review by innerthunder — APPROVED** [review]

**Pull request merged**